src/net/server/middleware/xfr/responder.rs

```diff
@@ -0,0 +1,161 @@
+use std::sync::Arc;
+
+use bytes::Bytes;
+use octseq::Octets;
+use tokio::sync::mpsc::{Receiver, UnboundedSender};
+use tokio::sync::Semaphore;
+use tracing::{debug, error};
+
+use crate::base::iana::OptRcode;
+use crate::base::rdata::RecordData;
+use crate::base::wire::Composer;
+use crate::base::{Message, Name, Rtype};
+use crate::net::server::batcher::ResourceRecordBatcher;
+use crate::net::server::middleware::xfr::util::add_to_stream;
+use crate::net::server::service::ServiceResult;
+use crate::net::server::util::mk_builder_for_target;
+use crate::zonetree::{Answer, SharedRrset};
+
+use super::batcher::{BatchReadyError, XfrRrBatcher};
+
+//------------ BatchingRrResponder ---------------------------------------------
+
+pub struct BatchingRrResponder<RequestOctets, Target> {
+    msg: Arc<Message<RequestOctets>>,
+    zone_soa_answer: Answer,
+    batcher_rx: Receiver<(Name<Bytes>, SharedRrset)>,
+    response_tx: UnboundedSender<ServiceResult<Target>>,
+    compatibility_mode: bool,
+    soft_byte_limit: usize,
+    must_fit_in_single_message: bool,
+    batcher_semaphore: Arc<Semaphore>,
+}
+
+impl<RequestOctets, Target> BatchingRrResponder<RequestOctets, Target>
+where
+    RequestOctets: Octets + Send + Sync + 'static + Unpin,
+    Target: Composer + Default + Send + Sync + 'static,
+{
+    #[allow(clippy::too_many_arguments)]
+    pub fn new(
+        msg: Arc<Message<RequestOctets>>,
+        zone_soa_answer: Answer,
+        batcher_rx: Receiver<(Name<Bytes>, SharedRrset)>,
+        response_tx: UnboundedSender<ServiceResult<Target>>,
+        compatibility_mode: bool,
+        soft_byte_limit: usize,
+        must_fit_in_single_message: bool,
+        batcher_semaphore: Arc<Semaphore>,
+    ) -> Self {
+        Self {
+            msg,
+            zone_soa_answer,
+            batcher_rx,
+            response_tx,
+            compatibility_mode,
+            soft_byte_limit,
+            must_fit_in_single_message,
+            batcher_semaphore,
+        }
+    }
+
+    pub async fn run(mut self) -> Result<(), OptRcode> {
+        // Limit the number of concurrently running XFR batching
+        // operations.
+        if self.batcher_semaphore.acquire().await.is_err() {
+            error!("Internal error: Failed to acquire XFR batcher semaphore");
+            return Err(OptRcode::SERVFAIL);
+        }
+
+        // SAFETY: msg.sole_question() was already checked in
+        // get_relevant_question().
+        let qclass = self.msg.sole_question().unwrap().qclass();
+
+        // Note: NSD apparently uses name compresson on AXFR responses
+        // because AXFR responses they typically contain lots of
+        // alphabetically ordered duplicate names which compress well. NSD
+        // limits AXFR responses to 16,383 RRs because DNS name
```

Comment:
Did you mean octets here?

Reply:
Yes. Thanks for spotting that. Replaced with bytes (as that is consistent with what is said further down in the same comment).